### PR TITLE
#18205 Prevent the endless growth of dbeaver-debug.log file

### LINKED ISSUE #18205: log file it's too big
### Description

November 11th, my finish work to home. but work computer not shutdown. November 14 th, computer remind me space is not enough.
截屏2022-11-14 上午10 05 04
截屏2022-11-14 上午10 05 35
截屏2022-11-14 上午10 05 46

### DBeaver Version

22.2.2.202210092258

### Operating System

macOS Bigsur 11.4

### Database and driver

mariadb

### Steps to reproduce

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by E1izabeth**
image

When the size of the log file exceeds the specified one (10Mb by default), the file is renamed (a timestamp is added) and then the writing continues to a new file. This happens the specified number of times - the maximum count of log files (3 by default).

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageErrorLogs.java [hunk 1]
```diff
@@ -71,6 +75,24 @@ protected Control createPreferenceContent(@NotNull Composite parent) {
...
                     GeneralUtils.variablePattern(SystemVariablesResolver.VAR_HOME)));
             logsDebugLocation.setLayoutData(new GridData(GridData.FILL_HORIZONTAL));
+            
+            UIUtils.createControlLabel(groupLogs, CoreMessages.pref_page_logs_files_max_size_label);
+            logFilesMaxSizeSpinner = new Spinner(groupLogs, SWT.BORDER);
+            logFilesMaxSizeSpinner.setDigits(0);
+            logFilesMaxSizeSpinner.setIncrement(10);
+            logFilesMaxSizeSpinner.setMinimum(0);
+            logFilesMaxSizeSpinner.setMaximum(Integer.MAX_VALUE);
+            long bigScriptSize = DBWorkbench.getPlatform().getPreferenceStore().getLong(LogOutputStream.LOGS_MAX_FILE_SIZE);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageErrorLogs.java]: Let's extract the common `DBWorkbench.getPlatform().getPreferenceStore()` part and inline these variables in other places

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageErrorLogs.java [hunk 2]
```diff
@@ -100,6 +128,9 @@ public boolean performOk()
         store.setValue(DBeaverPreferences.LOGS_DEBUG_ENABLED, logsDebugEnabled.getSelection());
         store.setValue(DBeaverPreferences.LOGS_DEBUG_LOCATION, logsDebugLocation.getText());
 
+        store.setValue(LogOutputStream.LOGS_MAX_FILE_SIZE, logFilesMaxSizeSpinner.getSelection() * 1024l);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/preferences/PrefPageErrorLogs.java]:
```suggestion
        store.setValue(LogOutputStream.LOGS_MAX_FILE_SIZE, logFilesMaxSizeSpinner.getSelection() * 1024L);
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/LogOutputStream.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+import org.jkiss.code.NotNull;
+import org.jkiss.dbeaver.model.preferences.DBPPreferenceListener;
+import org.jkiss.dbeaver.model.preferences.DBPPreferenceStore;
+import org.jkiss.dbeaver.runtime.DBWorkbench;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/LogOutputStream.java]: These should come before Java imports

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/LogOutputStream.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    private volatile int maxLogFiles;
+    
+    private final String logFileName;
+    private final String logFileNameExtension;
+    private final Predicate<String> logFileNamePattern;
+    
+    public LogOutputStream(@NotNull File debugLogFile) throws IOException {
+        if (debugLogFile.exists() && !debugLogFile.isFile()) {
+            throw new IOException(
+                "Failed to initialize debug log output due to the target is not a file: " + debugLogFile.getAbsolutePath()
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/LogOutputStream.java]:
```suggestion
                "Failed to initialize debug log output due to the target not being a file: " + debugLogFile.getAbsolutePath()
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/LogOutputStream.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            );
+        }
+        
+        this.currentLogFile = debugLogFile;
+        this.logFileLocation = debugLogFile.getParentFile();
+        this.prefStore = DBWorkbench.getPlatform().getPreferenceStore();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/LogOutputStream.java]: This field is not used outside the constructor. Can it be a local variable instead?

Or maybe you wanted to remove the property change listener from it in the `close` method?

File: plugins/org.jkiss.dbeaver.ui.app.standalone/src/org/jkiss/dbeaver/ui/app/standalone/DBeaverApplication.java [hunk 6]
```diff
@@ -687,14 +688,14 @@ private void initDebugWriter() {
...
-                System.err.println("Can't delete debug log file"); //$NON-NLS-1$
-                return;
-            }
-        }
+//        if (debugLogFile.exists()) {
+//            if (!debugLogFile.delete()) {
+//                System.err.println("Can't delete debug log file"); //$NON-NLS-1$
+//                return;
+//            }
+//        }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.app.standalone/src/org/jkiss/dbeaver/ui/app/standalone/DBeaverApplication.java]: Looks redundant now, can we delete it?

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Review by Destrolaric — APPROVED** [review]

**Comment by DariaMarkaryan** [comment]
Verified

**Comment by Matvey16** [comment]
Verified

**Pull request merged**